scipy/integrate/tests/test_tanhsinh.py

```diff
@@ -777,127 +778,136 @@ def f3(self, k, p):
+            assert xp.all((res.status == -1) & xp.isnan(res.sum)
+                          & xp.isnan(res.error) & ~res.success & res.nfev == 1)
 
         message = '...must be non-negative and finite.'
         with pytest.raises(ValueError, match=message):
-            nsum(f, f.a, f.b, tolerances=dict(rtol=-1))
+            nsum(f, a, b, tolerances=dict(rtol=-1))
         with pytest.raises(ValueError, match=message):
-            nsum(f, f.a, f.b, tolerances=dict(atol=np.inf))
+            nsum(f, a, b, tolerances=dict(atol=np.inf))
 
         message = '...may not be positive infinity.'
         with pytest.raises(ValueError, match=message):
-            nsum(f, f.a, f.b, tolerances=dict(rtol=np.inf), log=True)
+            nsum(f, a, b, tolerances=dict(rtol=np.inf), log=True)
         with pytest.raises(ValueError, match=message):
-            nsum(f, f.a, f.b, tolerances=dict(atol=np.inf), log=True)
+            nsum(f, a, b, tolerances=dict(atol=np.inf), log=True)
 
         message = '...must be a non-negative integer.'
         with pytest.raises(ValueError, match=message):
-            nsum(f, f.a, f.b, maxterms=3.5)
+            nsum(f, a, b, maxterms=3.5)
         with pytest.raises(ValueError, match=message):
-            nsum(f, f.a, f.b, maxterms=-2)
+            nsum(f, a, b, maxterms=-2)
 
     @pytest.mark.parametrize('f_number', range(1, 4))
-    def test_basic(self, f_number):
+    def test_basic(self, f_number, xp):
+        dtype = xp.asarray(1.).dtype
         f = getattr(self, f"f{f_number}")
-        res = nsum(f, f.a, f.b, args=f.args)
-        assert_allclose(res.sum, f.ref)
-        assert_equal(res.status, 0)
-        assert_equal(res.success, True)
+        a, b = xp.asarray(f.a), xp.asarray(f.b),
+        args = tuple(xp.asarray(arg) for arg in f.args)
+        ref = xp.asarray(f.ref, dtype=dtype)
+        res = nsum(f, a, b, args=args)
+        xp_assert_close(res.sum, ref)
+        xp_assert_equal(res.status, xp.zeros(ref.shape, dtype=xp.int32))
+        xp_test = array_namespace(a)  # CuPy doesn't have `bool`
+        xp_assert_equal(res.success, xp.ones(ref.shape, dtype=xp_test.bool))
 
         with np.errstate(divide='ignore'):
-            logres = nsum(lambda *args: np.log(f(*args)),
-                           f.a, f.b, log=True, args=f.args)
-        assert_allclose(np.exp(logres.sum), res.sum)
-        assert_allclose(np.exp(logres.error), res.error, atol=1e-15)
-        assert_equal(logres.status, 0)
-        assert_equal(logres.success, True)
-
+            logres = nsum(lambda *args: xp.log(f(*args)),
+                           a, b, log=True, args=args)
+        xp_assert_close(xp.exp(logres.sum), res.sum)
+        xp_assert_close(xp.exp(logres.error), res.error, atol=1e-15)
+        xp_assert_equal(logres.status, res.status)
+        xp_assert_equal(logres.success, res.success)
+
+    @pytest.mark.skip_xp_backends('torch', reason='seems sensitive to arithmetic')
```

Comment:
I wasn't very concerned because this test was very detail-oriented. I dug into it, though, and it passes with `xp.float64` in the right places. (The requested `rtol` is below the `eps` of `float32`.)